monai/handlers/utils.py

```diff
@@ -46,13 +46,24 @@ def stopping_fn(engine: Engine):
     return stopping_fn
 
 
-def all_gather(tensor):
+def evenly_divisible_all_gather(data: torch.Tensor) -> torch.Tensor:
     """
-    All gather the data of tensor value in distributed data parallel.
+    Utility function for distributed data parallel to pad at first dim to make it evenly divisible and all_gather.
+
+    Args:
+        data: source tensor to pad and execute all_gather in distributed data parallel.
+
     """
-    if not dist.is_available() or not dist.is_initialized():
-        raise RuntimeError("should not execute all_gather operation before torch.distributed is ready.")
-    # create placeholder to collect the data from all processes
-    output = [torch.zeros_like(tensor) for _ in range(dist.get_world_size())]
-    dist.all_gather(output, tensor)
-    return torch.cat(output, dim=0)
+    if idist.get_world_size() <= 1:
+        return data
+    # make sure the data is evenly-divisible on multi-GPUs
+    length = data.shape[0]
+    all_lens = idist.all_gather(length)
+    max_len = max(all_lens).item()
+    if length < max_len:
+        size = [max_len - length] + list(data.shape[1:])
+        data = torch.cat([data, data.new_full(size, float("NaN"))], dim=0)
```

Comment:
probably won't work for different data types, could you please add some distributed tests?

Reply:
Thanks for your suggestion, that's a good point, we don't need to use `float("NaN)` here actually.
Updated and added distributed tests in the latest commit.

Thanks.